### ISSUE TITLE
Change how postgres adapter inserts record for better concurrency support

### ISSUE THREAD

**Issue by jeremyevans** [jeremyevans/sequel#121]
Original here:

http://groups.google.com/group/datamapper/browse_thread/thread/90ec97f02c164e29

The CORRECT solution is pretty easy to implement, and would like to  
get feedback and possibly suggest to DO people.

```
If running PG 8.2, instead of INSERT;SELECT lastval(), do INSERT  
```

RETURNING id;  That will give you a concurrency safe way to do it,  
built into databases, etc.

```
The lastval() solution that is currently used I would ditch  
```

entirely, it's only an 8.1 solution (mine will work on ALL PG  
versions), and minimizes the race condition without eliminating.

```
If you are running PG 8.1 or earlier:
  id = SELECT nextval(sequence_name);
  INSERT INTO <table> (id, column1, column2, column3, etc.) VALUES  
```

(id, value1, value2, value3);

```
Basically, if you want the id back, instead of letting PostgreSQL  
```

set the id via the SERIAL, we're going to request the id, and then  
insert it with the id.  There is no down-side, it is the same two  
command to PG, we're incrementing the sequence in either case, etc,  
but we solve the race condition. 

Google Code Info:
Issue #: 96
Author: cico...@gmail.com
Created On: 2007-12-15T05:22:58.000Z
Closed On: 2008-02-05T08:20:16.000Z

**Comment by aaronpk** [comment]
Count me interested! Currently we're seeing a lot of work being done on the "currval" query because it requires a [full table scan](http://wordeology.com/computer/dont-use-postgress-currval.html).

Seems like Postgres <8.2 may not be so important to support anymore? Can we use the "RETURNING id" feature now?

**Comment by jeremyevans** [comment]
@aaronpk Sequel has used RETURNING id for many years by default on PostgreSQL, and dropped support for PostgreSQL < 8.2 years ago.  You are commenting on a ticket that was imported from Google Code, it was originally created back in 2007.

**Comment by aaronpk** [comment]
huh okay, so if I'm seeing queries using currval running right now am I using a ridiculously old version of the gem?

**Comment by jeremyevans** [comment]
Probably, or the queries are from another source, such as your code manually issuing them.  From a grep I don't believe Sequel uses `currval` internally on PostgreSQL at all, and I'm not sure if it ever did.  Old versions of Sequel would use `lastval` to get the inserted id, but support for that was dropped years ago.

**Comment by aaronpk** [comment]
Well here I was thinking I was being clever searching past issues instead of opening a new one.

You are correct, I found the place in our code where we are running currval. Unfortunately we had to add that because the gem doesn't know about partitioned tables. I will look for another solution to avoid running currval.
